### PR TITLE
Server support for headers between routes

### PR DESCRIPTION
**Pull request by derekcollison**
This also fixes a critical bug in the client headers implementation.

Signed-off-by: Derek Collison <derek@nats.io>

/cc @nats-io/core

### CODE REVIEW

File: test/new_routes_test.go [hunk 1]
```diff
@@ -50,6 +50,78 @@ func TestNewRouteInfoOnConnect(t *testing.T) {
...
+	sendA("PING\r\n")
+	expectA(pongRe)
+
+	if err := checkExpectedSubs(1, srvA, srvB); err != nil {
+		t.Fatalf("%v", err)
+	}
+
+	sendB, expectB := setupHeaderConn(t, clientB)
+	// Can not have \r\n in payload fyi for regex.
+	sendB("HPUB foo reply 12 14\r\nK1:V1,K2:V2 ok\r\n")
```
Comment (matthiashanel) [inline test/new_routes_test.go:81]: Maybe add tests without reply without a queue, without reply but a queue.

Comment (matthiashanel) [reply test/new_routes_test.go:81]: And maybe one that tests error conditions such as:
too few args, enough args but no + or | 
letters for a size for header and payload size.

File: server/route.go [hunk 2]
```diff
@@ -154,10 +155,91 @@ func (c *client) processAccountUnsub(arg []byte) {
...
+
+		// Grab queue names.
+		if c.pa.reply != nil {
+			c.pa.queues = args[4 : len(args)-1]
+		} else {
+			c.pa.queues = args[3 : len(args)-1]
+		}
+	}
+	if c.pa.size < 0 {
+		return fmt.Errorf("processRoutedHeaderMsgArgs Bad or Missing Size: '%s'", args)
```
Comment (matthiashanel) [inline server/route.go:229]: this covers a parsing error of the message body size.
shouldn't we have a similar thing for the header size?

Also make sure header size < msgsize holds true?

File: test/new_routes_test.go (hunk 1 above)
Comment (derekcollison) [reply test/new_routes_test.go:81]: I did those tests for HPUB, will add for HMSG too.

File: server/route.go (hunk 2 above)
Comment (derekcollison) [reply server/route.go:229]: Yep, did all of those for HPUB will add.

File: server/leafnode.go [hunk 3]
```diff
@@ -1442,6 +1442,11 @@ func (c *client) processLeafUnsub(arg []byte) error {
 	return nil
 }
 
+func (c *client) processLeafHeaderMsgArgs(arg []byte) error {
+	fmt.Printf("arg is %q\n", arg)
```
[the remark this answers is not in the document]
Comment (derekcollison) [reply server/leafnode.go:1446]: LOL yea..

[the remark this answers is not in the document]
Comment (derekcollison) [reply server/leafnode.go:1446]: To be implemented..

File: server/client.go [hunk 4]
```diff
@@ -2584,6 +2646,8 @@ func (c *client) deliverMsg(sub *subscription, subject, mh, msg []byte, gwrply b
 
 	// Check here if we have a header with our message. If this client can not
 	// support we need to strip the headers from the payload.
+	// The actual header would have been processed correctluy for us, so just
```
[the remark this answers is not in the document]
Comment (derekcollison) [reply server/client.go:2649]: fixed

File: server/leafnode.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (derekcollison) [reply server/leafnode.go:1446]: Eill just return error.

**Comment by derekcollison** [comment]
Please take another look when you have time.

File: server/parser.go [hunk 5]
```diff
@@ -1134,7 +1135,12 @@ func (c *client) clonePubArg() error {
 
 	switch c.kind {
 	case ROUTER, GATEWAY:
-		return c.processRoutedMsgArgs(c.argBuf)
+		if c.pa.hdr < 0 {
+			return c.processRoutedMsgArgs(c.argBuf)
+		} else {
+			fmt.Printf("ZZZZZ!!!\n")
```
Comment (matthiashanel) [inline server/parser.go:1141]: left over?

Comment (derekcollison) [reply server/parser.go]: removed and force pushed.

**Pull request merged**